reference-implementation/lib/writable-stream.js

```diff
@@ -204,44 +173,83 @@ function WritableStreamAddWriteRequest(stream) {
   return promise;
 }
 
-function WritableStreamFinishInFlightWrite(stream) {
-  assert(stream._inFlightWriteRequest !== undefined);
-  stream._inFlightWriteRequest._resolve(undefined);
-  stream._inFlightWriteRequest = undefined;
-
+function WritableStreamDealWithRejection(stream, error) {
   const state = stream._state;
 
-  if (state === 'errored') {
-    WritableStreamFinishInFlightWriteInErroredState(stream);
+  if (state === 'writable') {
+    WritableStreamErroring(stream, error);
     return;
   }
 
-  assert(state === 'writable');
+  assert(state === 'erroring');
+  WritableStreamFinishError(stream);
+}
+
+function WritableStreamErroring(stream, reason) {
+  assert(stream._storedError === undefined, 'stream._storedError === undefined');
+  assert(stream._state === 'writable', 'state must be writable');
+
+  const controller = stream._writableStreamController;
+  assert(controller !== undefined, 'controller must not be undefined');
 
-  WritableStreamHandleAbortRequestIfPending(stream);
+  stream._state = 'erroring';
+  stream._storedError = reason;
+  const writer = stream._writer;
+  if (writer !== undefined) {
+    WritableStreamDefaultWriterEnsureReadyPromiseRejected(writer, reason);
+  }
+
+  if (WritableStreamHasOperationMarkedInFlight(stream) === false && controller._started === true) {
+    WritableStreamFinishError(stream);
+  }
 }
 
-function WritableStreamFinishInFlightWriteInErroredState(stream) {
-  WritableStreamRejectAbortRequestIfPending(stream);
-  WritableStreamRejectPromisesInReactionToError(stream);
+function WritableStreamFinishError(stream) {
+  assert(stream._state === 'erroring', 'stream._state === erroring');
+  assert(WritableStreamHasOperationMarkedInFlight(stream) === false,
+         'WritableStreamHasOperationMarkedInFlight(stream) === false');
+  stream._state = 'errored';
+  stream._writableStreamController[ErrorSteps]();
+
+  const storedError = stream._storedError;
+  for (const writeRequest of stream._writeRequests) {
+    writeRequest._reject(storedError);
+  }
+  stream._writeRequests = [];
+
+  if (stream._pendingAbortRequest === undefined) {
```

Comment:
Could you please update `'controller.error(), writer.abort() while there is an in-flight write, and then finish the write'` test to check that `sink.abort()` is not called for this (error() then abort()) sequence?

controller.error() calls WritableStreamErroring() to change the state to erroring. WritableStreamAbort() skips WritableStreamErroring() when the state is erroring, but it does set _pendingAbortRequest. It should be rejected (right?) when controller.error() happened before writer.abort(), but it looks it would get processed by the code below.